charger/mcc.go

```diff
@@ -0,0 +1,406 @@
+		mcc.tokenValid = time.Now().Add(time.Duration(10) * time.Minute)
+
+		// the web interface updates the token every 2 minutes, so lets do the same here
+		mcc.tokenRefresh = time.Now().Add(time.Duration(2) * time.Minute)
+
+		return nil
+	}
+
+	return err
+}
+
+// login as the home user with the given password
+func (mcc *MCC) login(password string) error {
+	uri := fmt.Sprintf("%s%s", mcc.uri, apiLogin)
+
+	data := url.Values{
+		"user": []string{"user"},
+		"pass": []string{mcc.password},
+	}
+
+	req, err := http.NewRequest(http.MethodPost, uri, strings.NewReader(data.Encode()))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	err = mcc.fetchToken(req)
+
+	return err
+}
+
+// refresh the auth token with a new one
+func (mcc *MCC) refresh() error {
+	uri := fmt.Sprintf("%s%s", mcc.uri, apiRefresh)
+
+	req, err := http.NewRequest(http.MethodGet, uri, nil)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf(" Bearer %s", mcc.token))
+
+	err = mcc.fetchToken(req)
+
+	return err
+}
+
+// creates a custom http request that contains the auth token
+func (mcc *MCC) customRequest(method, uri string) (*http.Request, error) {
+	// do we need to login?
+	if mcc.token == "" || time.Since(mcc.tokenValid) > 0 {
+		if err := mcc.login(mcc.password); err != nil {
+			return nil, err
+		}
+	}
+
+	// is it time to refresh the token?
+	if time.Since(mcc.tokenRefresh) > 0 {
+		if err := mcc.refresh(); err != nil {
+			return nil, err
+		}
+	}
+
+	// now lets process the request with the fetched token
+	req, err := http.NewRequest(method, uri, nil)
+	if err != nil {
+		return req, err
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf(" Bearer %s", mcc.token))
+
+	return req, nil
+}
+
+// use http PUT to set a value on the URI, the value should be URL encoded in the URI parameter
+func (mcc *MCC) putValue(uri string) error {
+	req, err := mcc.customRequest(http.MethodPut, uri)
+	if err != nil {
+		return err
+	}
+
+	b, err := mcc.Request(req)
+	if err == nil {
+		var result string
+		err = json.Unmarshal(b, &result)
+
+		if err == nil {
+			if result != "OK" {
+				return fmt.Errorf("Call returned an unexpected error")
+			}
+		}
+	}
+
+	return err
+}
+
+// use http GET to fetch a non structured value from an URI and stores it in result
+func (mcc *MCC) getValue(uri string, result interface{}) error {
+	req, err := mcc.customRequest(http.MethodGet, uri)
+	if err != nil {
+		return err
+	}
+
+	b, err := mcc.Request(req)
+	if err == nil {
+		err = json.Unmarshal(b, &result)
+	}
+
+	return err
+}
+
+// use http GET to fetch an escaped JSON string and unmarshall the data in result
+func (mcc *MCC) getEscapedJSON(uri string, result interface{}) error {
+	req, err := mcc.customRequest(http.MethodGet, uri)
+	if err != nil {
+		return err
+	}
+
+	b, err := mcc.Request(req)
+	if err == nil {
+		var unquote string
+		if err := json.Unmarshal([]byte(b), &unquote); err != nil {
```

Comment:
wofür brauchst Du das doppelte unquote?

Reply:
Weil die API double encoded JSON zurückschickt 😢 

Beispiel:
```
"{\"carCable\":5,\"gridCable\":8,\"hwfpMaxLimit\":32,\"maxValue\":10,\"minValue\":6,\"value\":10}"
```

Reply:
Einfacher: https://stackoverflow.com/questions/16846553/how-to-unmarshal-an-escaped-json-string-in-go#16847249